sqld/src/replication/replica/injector/mod.rs

```diff
@@ -0,0 +1,220 @@
+                | OpenFlags::SQLITE_OPEN_NO_MUTEX,
+            &INJECTOR_METHODS,
+            // safety: hook is dropped after connection
+            unsafe { &mut *(ctx.as_mut() as *mut _) },
+            DEFAULT_AUTO_CHECKPOINT,
+        )?;
+
+        Ok(Self {
+            is_txn: false,
+            buffer,
+            capacity: buffer_capacity,
+            connection: Arc::new(Mutex::new(connection)),
+            _hook_ctx: ctx,
+        })
+    }
+
+    /// Inject on frame into the log. If this was a commit frame, returns Ok(Some(FrameNo)).
+    pub(crate) fn inject_frame(&mut self, frame: Frame) -> crate::Result<Option<FrameNo>> {
+        let frame_close_txn = frame.header().size_after != 0;
+        self.buffer.lock().push_back(frame);
+        if frame_close_txn || self.buffer.lock().len() >= self.capacity {
+            if !self.is_txn {
+                self.begin_txn()?;
+            }
+            return self.flush();
+        }
+
+        Ok(None)
+    }
+
+    /// Flush the buffer to libsql WAL.
+    /// Trigger a dummy write, and flush the cache to trigger a call to xFrame. The buffer's frame
+    /// are then injected into the wal.
+    fn flush(&mut self) -> crate::Result<Option<FrameNo>> {
+        let lock = self.buffer.lock();
+        // the frames in the buffer are either monotonically increasing (log) or decreasing
+        // (snapshot). Either way, we want to find the biggest frameno we're about to commit, and
+        // that is either the front or the back of the buffer
+        let last_frame_no = match lock.back().zip(lock.front()) {
+            Some((b, f)) => f.header().frame_no.max(b.header().frame_no),
+            None => {
+                tracing::trace!("nothing to inject");
+                return Ok(None);
+            }
+        };
+
+        drop(lock);
+
+        let connection = self.connection.lock();
+        connection.execute("INSERT INTO __DUMMY__ VALUES (42)", ())?;
+        // force call to xframe
+        match connection.cache_flush() {
+            Ok(_) => panic!("replication hook was not called"),
+            Err(e) => {
+                if let Some(e) = e.sqlite_error() {
+                    if e.extended_code == LIBSQL_INJECT_OK {
+                        // refresh schema
+                        connection.pragma_update(None, "writable_schema", "reset")?;
+                        let _ = connection.execute("COMMIT", ()); // TODO: Handle error?
+                        self.is_txn = false;
+                        assert!(self.buffer.lock().is_empty());
+                        return Ok(Some(last_frame_no));
+                    } else if e.extended_code == LIBSQL_INJECT_OK_TXN {
+                        self.is_txn = true;
+                        assert!(self.buffer.lock().is_empty());
+                        return Ok(None);
+                    } else if e.extended_code == LIBSQL_INJECT_FATAL {
+                        todo!("handle fatal error");
```

Comment:
Now those two TODOs make this PR a draft, right? We can't abort the program on injection error